src/frontend/src/webhook/mod.rs

```diff
@@ -0,0 +1,208 @@
+        Path((user, database, schema, table)): Path<(String, String, String, String)>,
+        body: Bytes,
+    ) -> Result<()> {
+        let session_mgr = SESSION_MANAGER
+            .get()
+            .expect("session manager has been initialized");
+
+        let dummy_addr = Address::Tcp(SocketAddr::new(
+            IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0)),
+            5691, // port of meta
+        ));
+
+        // TODO(kexiang): optimize this
+        // get a session object for the corresponding user and database
+        let session = session_mgr
+            .connect(database.as_str(), user.as_str(), Arc::new(dummy_addr))
+            .map_err(|e| {
+                err(
+                    anyhow!(e).context(format!(
+                        "Failed to create session for database: {}, user: {}",
+                        database, user
+                    )),
+                    StatusCode::UNAUTHORIZED,
+                )
+            })?;
+
+        let WebhookSourceInfo {
+            secret_ref,
+            signature_expr,
+        } = {
+            let search_path = session.config().search_path();
+            let user_name = user.as_str();
+            let schema_path = SchemaPath::new(Some(schema.as_str()), &search_path, user_name);
+
+            let reader = session.env().catalog_reader().read_guard();
+            let (table, _schema) = reader
+                .get_any_table_by_name(database.as_str(), schema_path, &table)
+                .map_err(|e| err(e, StatusCode::NOT_FOUND))?;
+
+            table
+                .webhook_info
+                .as_ref()
+                .ok_or_else(|| {
+                    err(
+                        anyhow!("Table {:?} is not connected to wehbook source", table),
+                        StatusCode::METHOD_NOT_ALLOWED,
+                    )
+                })?
+                .clone()
+        };
+
+        let secret_string = LocalSecretManager::global()
+            .fill_secret(secret_ref.unwrap())
+            .map_err(|e| err(e, StatusCode::NOT_FOUND))?;
+
+        // Once limitation here is that the key is no longer case-insensitive, users must user the lowercase key when defining the webhook source table.
+        let headers_jsonb = header_map_to_json(&headers);
+
+        let is_valid = verify_signature(
+            headers_jsonb,
+            secret_string.as_bytes(),
+            body.as_ref(),
+            signature_expr.unwrap(),
+            // signature,
+        )
+        .await?;
+
+        if !is_valid {
+            return Err(err(
+                anyhow!("Signature verification failed"),
+                StatusCode::UNAUTHORIZED,
+            ));
+        }
+
+        let payload = String::from_utf8(body.to_vec()).map_err(|e| {
+            err(
+                anyhow!(e).context("Failed to parse body"),
+                StatusCode::BAD_REQUEST,
+            )
+        })?;
+
+        let insert_stmt = Statement::Insert {
```

Comment:
> I guess the reason we put user in path is because you need a user to get his session and run the statement, right?

Mainly because I thought RW is a database. But since MZ doesn't require an user here, I removed the user field in the latest codes.

> First of all, I think the approach that building an insert statement and actually execute it, is somehow hacky. It may be better to share some logic in insert_handler.

Took a try before. Not very trivial. Will do in in the following patches.